hls-plugin-api/src/Ide/Types.hs

```diff
@@ -877,6 +886,57 @@ type CommandFunction ideState a
 
 -- ---------------------------------------------------------------------
 
+type ResolveFunction ideState a (m :: Method ClientToServer Request) =
+  ideState
+  -> PluginId
+  -> MessageParams m
+  -> Uri
+  -> a
+  -> LspM Config (Either ResponseError (MessageResult m))
+
+-- | Make a handler for plugins with no extra data
+mkResolveHandler
+  :: forall ideState a m. (FromJSON a,  PluginRequestMethod m, L.HasData_ (MessageParams m) (Maybe Value))
+  =>  SClientMethod m
+  -> (ideState
+  ->PluginId
+  -> MessageParams m
+  -> Uri
+  -> a
+  -> LspM Config (Either ResponseError (MessageResult m)))
+  -> PluginHandlers ideState
+mkResolveHandler m f = mkPluginHandler m f'
+  where f' ideState plId params = do
+          case fromJSON <$> (params ^. L.data_) of
+            (Just (Success (PluginResolveData owner uri value) )) -> do
+              if owner == plId
+              then
+                case fromJSON value of
+                  Success decodedValue ->
+                    let newParams = params & L.data_ ?~ value
+                    in f ideState plId newParams uri decodedValue
+                  Error err ->
+                    pure $ Left $ ResponseError (InR ErrorCodes_ParseError) (parseError value err) Nothing
+              else pure $ Left $ ResponseError (InR ErrorCodes_InvalidRequest) invalidRequest Nothing
+            _ -> pure $ Left $ ResponseError (InR ErrorCodes_InvalidRequest) invalidRequest Nothing
+        invalidRequest = "The resolve request incorrectly got routed to the wrong resolve handler!"
+        parseError value err = "Unable to decode: " <> (T.pack $ show value) <> ". Error: " <> (T.pack $ show err)
+
+wrapResolveData :: L.HasData_ a (Maybe Value) => PluginId -> Uri -> a -> a
+wrapResolveData pid uri hasData =
+  hasData & L.data_ .~  (toJSON .PluginResolveData pid uri <$> data_)
+  where data_ = hasData ^? L.data_ . _Just
+
+-- |Allow plugins to "own" resolve data, allowing only them to be queried for
+-- the resolve action. This design has added flexibility at the cost of nested
+-- Value types
```

Comment:
Do we actually need that any more? now thanks to `ResolveFunction` we know the inner `a`! So I think we could now make this polymorphic. That would mean we can decode the whole thing including the user value in one step, which would be nice.

Reply:
Hmm, but you put the ownership check in between the two steps, which probably makes sense. Otherwise we'll just get parse errors when we try to decode something that's for another plugin, which seems bad. So maybe this is the best way. worth writing something down